src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_drm.c

```diff
@@ -639,257 +659,209 @@ void xocl_drm_fini(struct xocl_drm *drm_p)
-				XOCL_IS_PS_KERNEL_MEM(grp_topology, i))
+		mem_data = &grp_topology->m_mem_data[i];
+		if ((convert_mem_tag(mem_data->m_tag) == MEM_TAG_HOST) ||
+				!XOCL_IS_PS_KERNEL_MEM(grp_topology, i))
 			continue;
 
-		hash_start_addr = grp_topology->m_mem_data[i].m_base_address;
-		hash_for_each_possible(drm_p->mm_range, wrapper, node, hash_start_addr) {
-			start_addr = wrapper->start_addr;
-			end_addr = wrapper->start_addr + wrapper->size;
+		start_addr = mem_data->m_base_address;
+		end_addr = start_addr + mem_data->m_size;
 
 #if defined(XOCL_DRM_FREE_MALLOC)
-			ret = drm_mm_insert_node_in_range(drm_p->mm, dnode, size, PAGE_SIZE, 0,
-					start_addr, end_addr, 0);
+		ret = drm_mm_insert_node_in_range(xocl_mm->mm, dnode, size, PAGE_SIZE, 0,
+				start_addr, end_addr, 0);
 #else
-			ret = drm_mm_insert_node_in_range(drm_p->mm, dnode, size, PAGE_SIZE,
-					start_addr, end_addr, 0);
+		ret = drm_mm_insert_node_in_range(xocl_mm->mm, dnode, size, PAGE_SIZE,
+				start_addr, end_addr, 0);
 #endif
-
-			if (!ret) 
-				/* Found the memory. Return it from here  */
-				return 0;
+		if (!ret) {
+			// Memory is allocated to this Bank
+			*mem_id = i;
+			return 0;
 		}
-	}	
-#endif
+	}
 
 	return ret;
 }
 
-
 static int xocl_mm_insert_node_range(struct xocl_drm *drm_p, u32 mem_id,
 		struct mem_topology *grp_topology, struct drm_mm_node *node, u64 size)
 {
+	struct xocl_mm *xocl_mm = drm_p->xocl_mm;
 	uint64_t start_addr = 0;
 	uint64_t end_addr = 0;
 	int ret = 0;
 
+	BUG_ON(!xocl_mm && !xocl_mm->mm);
 	start_addr = grp_topology->m_mem_data[mem_id].m_base_address;
 	end_addr = start_addr + grp_topology->m_mem_data[mem_id].m_size * 1024;
 
 #if defined(XOCL_DRM_FREE_MALLOC)
-	ret = drm_mm_insert_node_in_range(drm_p->mm, node, size, PAGE_SIZE, 0,
+	ret = drm_mm_insert_node_in_range(xocl_mm->mm, node, size, PAGE_SIZE, 0,
 					   start_addr, end_addr, 0);
 #else
-	ret = drm_mm_insert_node_in_range(drm_p->mm, node, size, PAGE_SIZE,
+	ret = drm_mm_insert_node_in_range(xocl_mm->mm, node, size, PAGE_SIZE,
 					   start_addr, end_addr, 0);
 #endif
 
 	return ret;
 }
 
-int xocl_mm_insert_node(struct xocl_drm *drm_p, u32 mem_id,
-			struct drm_mm_node *node, u64 size)
+int xocl_mm_insert_node(struct xocl_drm *drm_p, unsigned memidx,
+			uint32_t slotidx, struct drm_mm_node *node, u64 size)
 {
 	int ret = 0;
+	struct xocl_mem_stat *curr_mem = NULL;
 	struct mem_topology *grp_topology = NULL;
 	
 	BUG_ON(!mutex_is_locked(&drm_p->mm_lock));
-        if (drm_p->mm == NULL)
+        if (drm_p->xocl_mm->mm == NULL)
                 return -EINVAL;
 
 	ret = XOCL_GET_GROUP_TOPOLOGY(drm_p->xdev, grp_topology);
         if (ret)
                 return 0;
 
-	if (grp_topology->m_mem_data[mem_id].m_type == MEM_PS_KERNEL) {
-		ret = xocl_mm_insert_node_range_all(drm_p, mem_id, 
+	if (grp_topology->m_mem_data[memidx].m_type == MEM_PS_KERNEL) {
+		ret = xocl_mm_insert_node_range_all(drm_p, &memidx,
```

Comment:
Why is memidx changed inside insert? It does not look right.

Reply:
We need to send this for PS kernel. For PS kernel we need to find the available memory bank to allocate memory. 
Hence we need to pass memidx to get final bank where the memory has been allocated

Reply:
with line 777, it uses m_mem_data[memidx]. what if  778 returns different memidx?  Which one would be correct?

Reply:
As per PS kernel the memidx is dummy. 
             {
 20                 "m_type": "MEM_PS_KERNEL",
 21                 "m_used": "1",
 22                 "m_sizeKB": "0x0",
 23                 "m_tag": "MEM_PS_KERNEL",
 24                 "m_base_address": "0x0"
 25             }

It just tells that this is the request from a PS kernel. We need to create memory from actual BANKs. Hence memidx will be updated from where the actual memory has been allocated.

Reply:
Please add comment on this.